File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null && subRoot == null) return true;
        if(root == null && subRoot != null || root != null && subRoot == null) return false;
        if(root.val == subRoot.val) {
            if(subtreeHelper(root, subRoot)) {
                return true;
            }
        }
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }

    private boolean subtreeHelper(TreeNode p, TreeNode q) {
        if(p == null && q == null) return true;
        if(p != null && q != null && p.val == q.val) {
            return subtreeHelper(p.left, q.left) && subtreeHelper(p.right, q.right);
        } else {
            return false;
        }
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count = [0]
        def countNodes(root, maxVal, count):
            if root == None:
                return
            if root.val >= maxVal:
                count[0] += 1
            maxVal = max(maxVal, root.val)
            countNodes(root.left, maxVal, count)
            countNodes(root.right, maxVal, count)

        countNodes(root, root.val, count)
        return count[0]
